File: Hard Challenges/0001 - Reverse Guessing Game/Solution.py
class Guesser(object):
    def __init__(self, lower_bound=1, upper_bound=100):
        self.lower_bound = lower_bound
        self.upper_bound = upper_bound
        self.num_guesses = 0

    def check_validity(self):
        if self.lower_bound > self.upper_bound:
            print('Your number doesn\'t exist!')
            quit()

    def play_game(self):
        while True:
            self.check_validity()
            self.num_guesses += 1
            guess = self.guess_num()
            print('My guess is %d' % guess)
            answer = input('Is your number [h]igher, [l]ower, or was I [c]orrect?\n')
            if answer == 'h':
                self.lower_bound = guess + 1
            elif answer == 'l':
                self.upper_bound = guess - 1
            elif answer == 'c':
                print('I guessed your number in %d guess(es).' % self.num_guesses)
                quit()
            else:
                self.num_guesses -= 1
                print('Invalid input.')

    def guess_num(self):
        return int((self.upper_bound + self.lower_bound) / 2)


def main():
    number_guesser = Guesser()
    number_guesser.play_game()


if __name__ == '__main__':
    main()

File: Hard Challenges/0001 - Reverse Guessing Game/Solution.java
import java.util.Scanner;

public class Solution {

    private static Scanner reader = new Scanner(System.in);
    private static int numGuesses = 0;

    public static void main(String[] args) {
        guess(1,100);
    }

    public static void guess(int lowerBound, int upperBound) {
        numGuesses++;
        int guess = ((lowerBound + upperBound) / 2);
        System.out.format("My guess is %d.%n", guess);
        System.out.println("Is your number [h]igher, [l]ower, or was I [c]orrect?");
        switch (reader.next().charAt(0)) {
            case 'h':
                guess(guess + 1, upperBound);
                break;
            case 'l':
                guess(lowerBound, guess - 1);
                break;
            case 'c':
                System.out.printf("I guessed your number in %d guess(es).%n", numGuesses);
                break;
            default:
                numGuesses--;
                System.out.println("Invalid Input.");
                break;
        }

    }

}